File: data/snippets/quiz019.cpp
class Student {
private:
    std::string name;
    std::string id;
    std::vector<int> grades;
    
    // 允许其他类访问私有成员
    friend class CourseManager;
    friend class GradeCalculator;
    friend class SchoolReport;
    
public:
    Student(const std::string& name, const std::string& id) 
        : name(name), id(id) {}
    
    std::string getName() const { return name; }
    std::string getId() const { return id; }
    
    void addGrade(int grade) {
        grades.push_back(grade);
    }
};

class CourseManager {
public:
    void enrollStudent(Student& student, const std::string& courseCode) {
        std::cout << "Enrolling " << student.name << " in course " << courseCode << std::endl;
        // 直接访问并修改Student的私有属性
        student.id += "-" + courseCode;
    }
    
    void recordAttendance(Student& student, bool attended) {
        // 直接访问私有属性
        std::cout << "Recording attendance for " << student.name << " (" << student.id << ")" << std::endl;
    }
    
    void updateStudentGrades(Student& student, int newGrade) {
        // 直接操作私有数据
        student.grades.push_back(newGrade);
        std::cout << "Added grade " << newGrade << " for " << student.name << std::endl;
    }
};

class GradeCalculator {
public:
    double calculateAverage(const Student& student) {
        // 直接访问私有属性
        if (student.grades.empty()) {
            return 0.0;
        }
        
        int sum = 0;
        for (int grade : student.grades) {
            sum += grade;
        }
        
        return static_cast<double>(sum) / student.grades.size();
    }
    
    bool hasPassed(const Student& student, int passingGrade) {
        // 直接访问并使用私有数据
        for (int grade : student.grades) {
            if (grade < passingGrade) {
                return false;
            }
        }
        return !student.grades.empty();
    }
};

class SchoolReport {
public:
    void printReport(const Student& student) {
        // 直接访问私有属性
        std::cout << "Report for " << student.name << " (ID: " << student.id << ")" << std::endl;
        std::cout << "Grades: ";
        
        for (int grade : student.grades) {
            std::cout << grade << " ";
        }
        std::cout << std::endl;
    }
    
    void modifyStudentRecords(Student& student, const std::string& newName) {
        // 直接修改私有数据
        student.name = newName;
    }
};

File: data/snippets/quiz019.java
import java.util.List;
import java.util.ArrayList;

class Student {
    String name;
    String id;
    List<Integer> grades;
    
    public Student(String name, String id) {
        this.name = name;
        this.id = id;
        this.grades = new ArrayList<>();
    }
    
    public String getName() { return name; }
    public String getId() { return id; }
    
    public void addGrade(int grade) {
        grades.add(grade);
    }
}

class CourseManager {
    public void enrollStudent(Student student, String courseCode) {
        System.out.println("Enrolling " + student.name + " in course " + courseCode);
        student.id += "-" + courseCode;
    }
    
    public void updateStudentGrades(Student student, int newGrade) {
        student.grades.add(newGrade);
        System.out.println("Added grade " + newGrade + " for " + student.name);
    }
}

class GradeCalculator {
    public double calculateAverage(Student student) {
        if (student.grades.isEmpty()) {
            return 0.0;
        }
        
        int sum = 0;
        for (int grade : student.grades) {
            sum += grade;
        }
        
        return (double) sum / student.grades.size();
    }
    
    public boolean hasPassed(Student student, int passingGrade) {
        for (int grade : student.grades) {
            if (grade < passingGrade) {
                return false;
            }
        }
        return !student.grades.isEmpty();
    }
}

class SchoolReport {
    public void printReport(Student student) {
        System.out.println("Report for " + student.name + " (ID: " + student.id + ")");
        System.out.print("Grades: ");
        
        for (int grade : student.grades) {
            System.out.print(grade + " ");
        }
        System.out.println();
    }
    
    public void modifyStudentRecords(Student student, String newName) {
        student.name = newName;
    }
}
